File: shops/electrosila.py
import sys
from urllib.parse import urlencode

import requests
import collections
from bs4 import BeautifulSoup

from entities import Product
from shops.shop import Shop

collections.Callable = collections.abc.Callable


class ElectroSila(Shop):
    title = "ElectroSila"

    def find(self, query):
        soup = self.send_request(query)

        products = []
        products_html = soup.find_all("div", attrs={"class": "tov_prew_search"})
        for product_html in products_html:
            result_link = product_html.find_next("a")
            if not result_link:
                continue
            result_link = result_link.attrs.get("href")

            result_name = product_html.find_next("strong")
            if not result_name:
                continue
            result_name = result_name.text

            result_price_block = product_html.find_next("div", attrs={"class": "price"})
            if not result_price_block:
                continue

            result_prices = result_price_block.find_next("div")
            if not result_prices:
                continue

            result_price = result_prices.find_all("b")[:2]
            if not result_price:
                continue

            price_string = ""
            i = 1
            for tag in result_price:
                price_string += tag.text.replace(".", "")
                if i == 1:
                    price_string += "."
                i += 1

            product = Product(self.title)
            product.title = result_name
            product.price = float(price_string)
            product.link = result_link

            products.append(product)

        return products

    @staticmethod
    def send_request(query):
        url = 'https://sila.by/search/%s/sort/6' % query
        response = requests.get(url, headers={
            "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 "
                          "Safari/537.36"
        })
        soup = BeautifulSoup(response.text, 'html.parser')

        return soup

File: pkg/shopsparser/shops/electrosila.go
package shops

import (
	"belby/pkg/request"
	"fmt"
	"golang.org/x/net/html"
	"golang.org/x/text/encoding/charmap"
	"io"
	"net/http"
	"sort"
	"strconv"
	"strings"
)

type ElectroSila struct {
	title string
}

func NewElectroSilaShop() *ElectroSila {
	return &ElectroSila{
		title: "ЭлектроСила",
	}
}

func (s *ElectroSila) GetTitle() string {
	return s.title
}

func (s *ElectroSila) Find(query string) (Product, error) {
	var products []Product

	response, err := s.makeRequest(query)
	if err != nil {
		return Product{}, err
	}

	products, err = s.parse(response)
	if err != nil {
		return Product{}, nil
	}

	sortedProducts, err := s.sort(products)
	if err != nil {
		return Product{}, nil
	}

	if len(sortedProducts) > 0 {
		return sortedProducts[0], nil
	} else {
		return Product{}, nil
	}
}

func (s *ElectroSila) sort(products []Product) ([]Product, error) {
	sortedProducts := products

	sort.Slice(sortedProducts[:], func(i, j int) bool {
		return sortedProducts[i].Price < sortedProducts[j].Price
	})

	return sortedProducts, nil
}

func (s *ElectroSila) parse(response string) ([]Product, error) {
	var products []Product

	tokenizer := html.NewTokenizer(strings.NewReader(response))

	divFound := false
	divsCountBeforeClose := 0
	titleFound := false
	priceFound := false
	priceDecimalFound := 0

	product := Product{
		ShopTitle: s.GetTitle(),
	}
	for {
		tn := tokenizer.Next()

		switch {
		case tn == html.ErrorToken:
			return products, nil
		case tn == html.TextToken:
			if divFound {
				token := tokenizer.Token()
				if titleFound {
					product.Title = token.Data
					titleFound = false
				}
				if priceDecimalFound != 0 {
					priceDecimalFound++
				}
				if priceDecimalFound == 3 {
					decimals, err := strconv.ParseFloat(token.Data, 64)
					if err != nil {
						continue
					}
					product.Price += decimals

					priceDecimalFound = 0

					products = append(products, product)
					product = Product{
						ShopTitle: s.GetTitle(),
					}
				}
				if priceFound {
					priceFound = false

					price, err := strconv.ParseFloat(token.Data, 64)
					if err != nil {
						continue
					}
					product.Price = price
					priceDecimalFound = 1
				}
			}
		case tn == html.StartTagToken:
			token := tokenizer.Token()

			if divFound && token.Data == "div" {
				for _, attribute := range token.Attr {
					if attribute.Key == "class" && attribute.Val == "price" {
						priceFound = true
					}
				}
				divsCountBeforeClose++
			}

			if divFound && token.Data == "strong" {
				titleFound = true
			}

			if divFound && token.Data == "a" && product.Link == "" {
				for _, attribute := range token.Attr {
					if attribute.Key == "href" {
						product.Link = attribute.Val
						break
					}
				}
			}

			for _, attribute := range token.Attr {
				if attribute.Key == "class" && attribute.Val == "tov_prew_search" {
					divFound = true
					divsCountBeforeClose = 1
				}
			}
		case tn == html.EndTagToken:
			if divFound {
				token := tokenizer.Token()
				if token.Data == "div" {
					divsCountBeforeClose--

					if divsCountBeforeClose == 0 {
						divFound = false
					}
				}
			}
		}
	}
}

func (s *ElectroSila) makeRequest(query string) (string, error) {
	body, err := request.SendRequest(request.Request{
		Method: http.MethodGet,
		URL:    fmt.Sprintf("https://sila.by/search/%s/sort/6", query),
		Headers: map[string]string{
			"User-Agent": DefaultUserAgent,
		},
		Callback: func(body io.ReadCloser) ([]byte, error) {
			reader := charmap.Windows1251.NewDecoder().Reader(io.Reader(body))
			return io.ReadAll(reader)
		},
	})
	if err != nil {
		return "", err
	}

	return string(body), nil
}
